src/core/fs.rs

```diff
@@ -117,3 +121,246 @@ pub fn set_group(path: &Path, group: &str) -> Result<(), io::Error> {
+    // SAFETY: always safe to call
+    unsafe { *libc::__errno_location() }
+}
+
+fn set_fd_nonblocking(fd: RawFd) -> Result<(), io::Error> {
+    // SAFETY: always safe to call
+    let flags = unsafe { libc::fcntl(fd, libc::F_GETFL, 0) };
+    if flags < 0 {
+        return Err(io::Error::last_os_error());
+    }
+
+    // SAFETY: always safe to call
+    let ret = unsafe { libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK) };
+    if ret != 0 {
+        return Err(io::Error::last_os_error());
+    }
+
+    Ok(())
+}
+
+struct FileWatcherState {
+    watcher: notify::RecommendedWatcher,
+    changed: bool,
+}
+
+struct FileWatcherData {
+    file: PathBuf,
+    read_fd: RawFd,
+    write_fd: RawFd,
+    state: Mutex<Option<FileWatcherState>>,
+}
+
+pub struct FileWatcher {
+    data: Arc<FileWatcherData>,
+}
+
+#[derive(Debug)]
+pub struct FileWatcherError;
+
+impl FileWatcher {
+    pub fn new<P: AsRef<Path>>(file_path: P) -> Result<Self, FileWatcherError> {
+        let file = file_path.as_ref();
+
+        let dir = match file.parent() {
+            Some(p) => p,
+            None => return Err(FileWatcherError),
+        };
+
+        let mut fds = [0; 2];
+
+        // SAFETY: fds pointer is valid
+        let ret = unsafe { libc::pipe(fds.as_mut_ptr()) };
+        assert_eq!(ret, 0);
+
+        for fd in &fds {
+            set_fd_nonblocking(*fd).unwrap();
+        }
+
+        let data = Arc::new(FileWatcherData {
+            file: file.to_owned(),
+            read_fd: fds[0],
+            write_fd: fds[1],
+            state: Mutex::new(None),
+        });
+
+        let watcher = {
+            let data = Arc::clone(&data);
+
+            notify::recommended_watcher(move |event: Result<notify::Event, notify::Error>| {
+                let event = match event {
+                    Ok(event) => event,
+                    Err(e) => {
+                        warn!("file watcher error: {:?}", e);
+                        return;
+                    }
+                };
+
+                if !event.paths.into_iter().any(|p| p == data.file) {
+                    // skip unrelated events
+                    return;
+                }
+
+                if let Some(state) = data.state.lock().unwrap().as_mut() {
+                    if !state.changed {
+                        state.changed = true;
+
+                        // non-blocking write to wake up the other side
+
+                        let buf: [u8; 1] = [0; 1];
+
+                        // SAFETY: buf pointer and size are valid
+                        let ret = unsafe {
+                            libc::write(data.write_fd, buf.as_ptr() as *const libc::c_void, 1)
+                        };
+                        assert!(ret == 1 || get_errno() == libc::EAGAIN);
+                    }
+                }
+            })
+            .expect("failed to create file watcher")
+        };
+
+        {
+            let mut state = data.state.lock().unwrap();
```

Comment:
Same Nitpick: state lock unwrap that should at least be an .expect() to help if this ever goes wrong.